Deep Search: updates for 6.5 (#1191)
sourcegraph/docs: docs/code-search/types/deep-search.mdx

# Deep Search

<p className= "subtitle">Learn more about Sourcegraph's agentic Code Search tool Deep Search.</p>

<Callout type="note"> New in version 6.5. Deep Search is currently in research preview for Enterprise and Enterprise Starter customers. Because Deep Search is in research preview, it might change significantly as we improve and adjust to user feedback. Please reach out to your Sourcegraph account team to request access. </Callout>

Deep Search is an agentic code search tool that understands natural language questions about your codebase. When a question is submitted, Deep Search performs an in-depth search and returns a detailed answer. The conversation can be continued with follow-up questions to dive deeper into relevant code.

> Under the hood, Deep Search is an AI agent that uses various tools to generate its answer. The tools are functionalities available in Sourcegraph. They include multiple modes of Sourcegraph's Code Search and Code Navigation features. All processing happens within your Sourcegraph instance. Only external calls are made to the configured LLM.

Author: @MaedahBatool  Could you please change the last sentence to `The only external calls Deep Search makes are to the configured LLM`  - unfortunately your edit changed the meaning.

Every Deep Search response includes a detailed list of sources contributing to the answer. These sources show exactly which searches were performed and which files were read. The list of sources is extremely useful for understanding where the answer came from and for further explorations of the codebase.

The answer is formatted in Markdown and can include links to relevant files, directories, or repositories. If prompted to do so, Deep Search can also generate diagrams as part of its answer.

## Best practices

- Give the agent a starting point for the search: use @-mentions to mention relevant repositories, files, directories, or symbol names. The more specific you are, the faster the search will be.
- Provide reasonably scoped questions. The agent will perform much better if it does not have to read the entire codebase at once.
- Check the list of sources. This is extremely useful for debugging and understanding where the answer came from. Ask a follow-up question and mention the missing source if something is missing.

> For use cases where you're looking for exhaustive answers (for example, "Find all files with the `.XYZ` file extension in `foo' repo that contain the word `bar`), Code Search still excels, while Deep Search will only utilize a sample of the results. Deep Search will perform a Code Search query as a source, which you can use to continue an exhaustive search within the Code Search product.

Author: @MaedahBatool Please keep the `continue with an exhaustive search` part - unfortunately your edit changes the meaning a bit which isn't what we want.

<p className="subtitle">Learn more about Sourcegraph's agentic Code Search tool Deep Search.</p>

> <Callout type="note"> New in version 6.4. Deep Search is currently in research preview for Enterprise customers. Because Deep Search is in research preview, it might change significantly in the future as we make improvements and adjust to user feedback. Please reach out to your Sourcegraph account team to request access. </Callout>

Reviewer: Should this say 6.5? Yes we released to some customers in 6.4 but this broader release is for 6.5

Author: Fair enough - we're moving really fast here so this serving as an additional nudge to upgrade makes a lot of sense.

- Look at the GraphQL APIs available in `@sourcegraph/sourcegraph`. Are any of them unused? The client code is in the `@cody` repository.
- Which tools do we use in our build processes defined in `BUILD.bazel` files?
- Generate a request flow diagram for `src/backend`. Mark the auth and rate limit points.

## Conversation sharing
You can share Deep Search conversations with other users in your Sourcegraph instance. To share a conversation, click the "Share" button in the top left, then copy the link. 
Once you share a conversation, any user on your instance is able to view it with the link. You can also reset the share link and generate a new one, which invalidates the previous link.

Note that currently, we do not enforce [permissions](/admin/permissions) for viewing shared Deep Search conversations. This means that any user in your Sourcegraph instance can view any shared conversation, regardless of which repositories they have access to. We plan to add support for permission checking in the future.

## Enabling Deep Search

> Deep Search is disabled by default. To enable it, ask your site administrator to enable the following setting in your site configuration:

Reviewer 2: Yea and we can still say if it is disabled you contact your site administrator

<p className="subtitle">Learn more about Sourcegraph's agentic Code Search tool Deep Search.</p>

> <Callout type="note"> New in version 6.5. Deep Search is currently in research preview for Enterprise customers. Because Deep Search is in research preview, it might change significantly in the future as we make improvements and adjust to user feedback. Please reach out to your Sourcegraph account team to request access. </Callout>

Reviewer 2: Enterprise and Enterprise Starter

Deep Search is an agentic code search tool that understands natural language questions about your codebase. When a question is submitted, Deep Search performs an in-depth search and returns a detailed answer. The conversation can be continued with follow-up questions to dive deeper into relevant code.

> Under the hood, Deep Search is an AI agent that uses various tools to generate its answer. The tools are functionalities available in Sourcegraph. They include multiple modes of Sourcegraph's Code Search and Code Navigation features. All processing happens within your Sourcegraph instance - only LLM calls are made to external providers based on your configuration.

Reviewer 2: > only LLM calls are made to external providers based on your configuration

That reads like we ONLY make LLM calls, but I think you intend to say "the only LLM calls made are based on your configuration". But even that is a bit confusing since Cloud customers don't configure the LLM (only self-hosted). 

Can we even just drop this line?

Author: I put this in because we mentioned something once about having an idea of the data flow for DS - basically what happens inside it. I can remove it, but do we want to include a variant of this somewhere else?

### Examples of prompts

- Find examples of logger usage and show examples of the different types of logging we use.
- I want to know when the indexing queue functionality was last changed in `@zoekt`. Show me the last few commit diffs touching this code and explain the changes.
- Look at the GraphQL APIs available in `@sourcegraph/sourcegraph`. Are any of them unused? The client code is in the `@cody` repository.
- Which tools do we use in our build processes defined in `BUILD.bazel` files?
- Generate a request flow diagram for `src/backend`. Mark the auth and rate limit points.

## Conversation sharing
You can share Deep Search conversations with other users in your Sourcegraph instance. To share a conversation, click the "Share" button in the top left, then copy the link. 
Once you share a conversation, any user on your instance is able to view it with the link. You can also reset the share link and generate a new one, which invalidates the previous link.

> Note that currently, we do not enforce [permissions](/admin/permissions) for viewing shared Deep Search conversations. This means that any user in your Sourcegraph instance can view any shared conversation, regardless of which repositories they have access to. We plan to add support for permission checking in the future.

Reviewer 2: Should this be a call out right below the `## Conversation sharing` header?

> Note that currently, we do not enforce [permissions](/admin/permissions) for viewing shared Deep Search conversations. This means that any user in your Sourcegraph instance can view any shared conversation, regardless of which repositories they have access to. We plan to add support for permission checking in the future.

Reviewer 2: ```suggestion
Currently, we do not enforce [repository permissions](/admin/permissions) for viewing shared Deep Search conversations. This means that a user can view a conversation shared to them, regardless of which repositories they have access to. We plan to revisit this in the future.
```

## Enabling Deep Search

If Deep Search is disabled, ask your site administrator to enable the following setting in your site configuration:
```json
"experimentalFeatures": {
    "deepSearch.enabled": true,
},
```

For optimal performance, Deep Search is specialized to only use one model. Currently, Deep Search only supports Claude Sonnet 4.

If you are not using [Cody Gateway](/cody/core-concepts/cody-gateway), you will also need to configure the model through either Amazon Bedrock or GCP Vertex.

> ### Enabling conversation sharing

Reviewer 2: Should we also mention its disabled by default in the section higher up when you mention sharing?
